src/Voron/Impl/WeakSliceSmallSet.cs

```diff
@@ -0,0 +1,142 @@
+using Sparrow.Server;
+
+namespace Voron.Impl
+{
+    // The concept of the weak small set is about an optimized set that can lose elements when it gets filled.
+    // In a sense it behaves like a LRU cache, but will not track the read accesses. This version of the SmallSet
+    // is specially designed to deal with sequence values stored in byte pointers. The main difference with a
+    // dictionary is that the small set will calculate the hash ONLY if there are candidates that match already. 
+    public sealed class WeakSliceSmallSet<TValue> : IDisposable
+    {
+        private const int Invalid = -1;
+
+        private readonly int _length;
+        private readonly int[] _keySizes;
+        private readonly ulong[] _keyHashes;
+        private readonly Slice[] _keys;
+        private readonly TValue[] _values;
+        private int _currentIdx;
+
+        public WeakSliceSmallSet(int size = 0)
+        {
+            _length = size > Vector<long>.Count ? size - size % Vector<long>.Count : Vector<long>.Count;
+            _keySizes = ArrayPool<int>.Shared.Rent(_length);
+            _keyHashes = ArrayPool<ulong>.Shared.Rent(_length);
+            _keys = ArrayPool<Slice>.Shared.Rent(_length);
+            _values = ArrayPool<TValue>.Shared.Rent(_length);
+            _currentIdx = -1;
+        }
+
+        public unsafe void Add(Slice key, TValue value)
+        {
+            int idx = FindKey(key);
+            if (idx == Invalid)
+                idx = RequestWritableBucket();
+
+            _keys[idx] = key;
+            _keySizes[idx] = key.Size;
+            _keyHashes[idx] = Hashing.XXHash64.CalculateInline(key.Content.Ptr, (ulong)key.Size);
+            _values[idx] = value;
+
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private unsafe int FindKey(Slice key)
+        {
+            if (_currentIdx == Invalid)
+                return Invalid;
+
+            int keyLength = key.Size;
+            ulong keyHash = 0;
+            
+            Slice[] keys = _keys;
+            int[] keySizes = _keySizes;
+            ulong[] keyHashes = _keyHashes;
+            byte* keyPtr = key.Content.Ptr;
+
+            // PERF: It may seems strange to increase the size to decrement it as the first
+            // loop operation. The reason behind this is to be able to just jump back immediately
+            // to the top of the loop as soon as we know the item is not the item we are looking
+            // for. 
+            int elementIdx = Math.Min(_currentIdx, _length - 1);
+            while (elementIdx >= 0)
+            {
+                var currentIdx = elementIdx;
+                elementIdx--;
+
+                // First check, we are not going to look into any string that is not of the correct size
+                if (keySizes[currentIdx] != keyLength)
+                    continue;
+
+                // PERF: Assuming a uniformly random symbol distribution, the chance that first two symbols match
```

Comment:
Assume that we have a lot of collections, named:

`UserPreferences`, `UserLogs`, `UserMatches`, etc.

So it is likely to have a lot of collision on the prefixes and suffices.

Reply:
`Lua` does a check on first, last and mid chars, which isn't always good, but may be better option

Reply:
Yes, agree. Will include that.